File: src/falcon_core/math/spaces/cartesian_2D_space.py
"""A Cartesian 2D space."""

from .cartesian_space import CartesianSpace
from .dependancies import Domain


class Cartesian2DSpace(CartesianSpace):
    """A Cartesian 2D space.

    A Cartesian 2D space is a unit space with a two axis.
    """

    def __init__(
        self,
        deltas: list[float],
        domain=Domain(bounds=(0, 1), greater_bound_contained=False),
    ):
        """Initialize the Cartesian1DSpace object."""
        assert len(deltas) == 2, "A 2D space must have two deltas."
        super().__init__(
            deltas=deltas,
            domain=domain,
        )

File: src/math/spaces/Cartesian2DSpace.cpp
#include "falcon_core/math/spaces/Cartesian2DSpace.hpp"

namespace falcon_core::math::spaces {

Cartesian2DSpace::Cartesian2DSpace() = default;
Cartesian2DSpace::Cartesian2DSpace(const std::vector<double>& deltas,
                                   const domains::DomainSP&   domain)
    : CartesianSpace(deltas, domain) {
  if (deltas.size() != 2) {
    throw std::invalid_argument("A 2D space must have two deltas.");
  }
}

}  // namespace falcon_core::math::spaces

CEREAL_REGISTER_TYPE(falcon_core::math::spaces::Cartesian2DSpace)
CEREAL_REGISTER_POLYMORPHIC_RELATION(
    falcon_core::generic::Song, falcon_core::math::spaces::Cartesian2DSpace)
